stage2.1_traited_script/traited_face_detect.py

```diff
@@ -0,0 +1,90 @@
+""" Script analyzing an image, detecting human faces inside it, and printing
+EXIF data about it.
+"""
+import PIL.Image
+from PIL.ExifTags import TAGS
+from skimage import data
+from skimage.feature import Cascade
+import matplotlib.pyplot as plt
+from matplotlib import patches
+from os.path import join
+import numpy as np
+
+# ETS imports
+from traits.api import Dict, File, HasStrictTraits, List, observe
+
+
+class ImageFile(HasStrictTraits):
+    """ Model to hold an image file.
+    """
+    filepath = File
+
+    faces = List
+
+    metadata = Dict
+
+    def to_array(self):
+        with PIL.Image.open(self.filepath) as img:
+            return np.asarray(img)
+
+    @observe("filepath")
+    def update_metadata(self, event):
+
+        with PIL.Image.open(self.filepath) as img:
+            exif = img._getexif()
+        self.metadata = {TAGS[k]: v for k, v in exif.items()
+                         if k in TAGS}
+
+    def detect_faces(self):
+        # Load the trained file from the module root.
+        trained_file = data.lbp_frontal_face_cascade_filename()
+
+        # Initialize the detector cascade.
+        detector = Cascade(trained_file)
+
+        detected = detector.detect_multi_scale(img=self.to_array(),
+                                               scale_factor=1.2,
+                                               step_ratio=1,
+                                               min_size=(60, 60),
+                                               max_size=(600, 600))
+        self.faces = detected
+
+        self.metadata["Number of faces detected"] = len(detected)
```

Comment:
This may be intentional, and feel free to correct me, but the user must call `detect_faces` immediately after `filepath` is modified or else the `ImageFile` model can get out of sync (`image.faces` could correspond to the previous `filepath`).

One way to fix that would be to set up a listener on `filepath`, but that isn't always desirable either. Sometimes we  want to compute properties lazily. Ideally, `faces` would be a `Property`, but this is a tutorial and we want to introduce concepts gently.

As a compromise, what do you think about not storing `faces` on the model and adding a `detect_faces` helper function which returns `faces` instead?
```python
image_file.filepath = "/path/to/file.png"
faces = detect_faces(image_file)
image_file.metadata["Num faces"] = len(faces)
```

Reply:
I agree with you, this is a perfect situation for introducing properties to illustrate the lazy loading situation. I could work on a 2.1 version that adds the property concept and have this discussion.

Reply:
I agree with you, this is a perfect situation for introducing properties to illustrate the lazy loading situation. I could work on a 2.1 version that adds the property concept and have this discussion.

Reply:
That sounds great!